active-rfcs/0000-function-api.md

```diff
@@ -0,0 +1,643 @@
+  }
+}
+```
+
+### Effect Cleanup
+
+Sometimes the watcher callback will perform async side effects that need to be invalidated when the watched value changes. The watcher callback receives a 3rd argument that can be used to register a cleanup function. The cleanup function is called when:
+
+- the watcher is about to re-run
+- the watcher is stopped (i.e. when the component is unmounted if `watch` is used inside `setup()`)
+
+``` js
+watch(idValue, (id, oldId, onCleanup) => {
+  const token = performAsyncOperation(id)
+  onCleanup(() => {
+    // id has changed or watcher is stopped.
+    // invalidate previously pending async operation
+    token.cancel()
+  })
+})
+```
+
+We are registering cleanup via a passed-in function instead of returning it from the callback (like React `useEffect`) because the return value is important for async error handling. It is very common for the watcher callback to be an async function when performing data fetching:
+
+``` js
+const data = value(null)
+watch(getId, async (id) => {
+  data.value = await fetchData(id)
+})
+```
+
+An async function implicitly returns a Promise, but the cleanup function needs to be registered immediately before the Promise resolves. In addition, Vue relies on the returned Promise to automatically handle potential errors in the Promise chain.
+
+### Watcher Callback Timing
+
+By default, all watcher callbacks are fired **after current renderer flush.** This ensures that when callbacks are fired, the DOM will be in already-updated state. If you want a watcher callback to fire before flush or synchronously, you can use the `flush` option:
+
+``` js
+watch(
+  () => count.value + 1,
+  () => console.log(`count changed`),
+  {
+    flush: 'post', // default, fire after renderer flush
+    flush: 'pre', // fire right before renderer flush
+    flush: 'sync' // fire synchronously
+  }
+)
+```
+
+### Full `watch` Options
+
+``` ts
+interface WatchOptions {
+  lazy?: boolean
+  deep?: boolean
+  flush?: 'pre' | 'post' | 'sync'
+  onTrack?: (e: DebuggerEvent) => void
+  onTrigger?: (e: DebuggerEvent) => void
+}
+
+interface DebuggerEvent {
+  effect: ReactiveEffect
+  target: any
+  key: string | symbol | undefined
+  type: 'set' | 'add' | 'delete' | 'clear' | 'get' | 'has' | 'iterate'
+}
+```
+
+- `lazy` is the opposite of 2.x's `immediate` option.
+- `deep` works the same as 2.x
+- `onTrack` and `onTrigger` are hooks that will be called when a dependency is tracked or the watcher getter is triggered. They receive a debugger event that contains information on the operation that caused the track / trigger.
+
+## Lifecycle Hooks
+
+All current lifecycle hooks will have an equivalent `onXXX` function that can be used inside `setup()`:
+
+``` js
+import { onMounted, onUpdated, onUnmounted } from 'vue'
+
+const MyComponent = {
+  setup() {
+    onMounted(() => {
+      console.log('mounted!')
+    })
+    onUpdated(() => {
+      console.log('updated!')
+    })
+    onUnmounted(() => {
```

Comment:
Is the renaming from *destroy* to *unmount* verb intentional?

Or is the *mount*/*unmount* pair analogous to *create*/*destroy* pair?

Reply:
Yes, `destroy` -> `unmount` is intentional.

Reply:
> All current lifecycle hooks will have an equivalent onXXX function that can be used inside setup()  [stated here](https://github.com/vuejs/rfcs/blob/function-apis/active-rfcs/0000-function-api.md#lifecycle-hooks)

Since there will be a renaming, I'd recommend list all available lifecycle hook functions in the proposal explicitly.